lib/pure/collections/lists.nim

```diff
@@ -458,20 +458,32 @@ proc prepend*[T](L: var SinglyLinkedList[T], value: T) {.inline.} =
     assert a.contains(9)
   prepend(L, newSinglyLinkedNode(value))
 
-proc add*[T](L1: var SinglyLinkedList[T], L2: SinglyLinkedList[T]) {.since: (1, 5).} =
-  ## Adds `L2` to the end of `L1`. Efficiency: O(1).
-  ## Note that the two lists share structure after the operation.
+func copy*[T](L: SinglyLinkedList[T]): SinglyLinkedList[T] {.since: (1, 5).} =
+  ## Creates a shallow copy of `L`.
   runnableExamples:
-    var a = [1, 2, 3].toSinglyLinkedList
-    let b = [4, 5].toSinglyLinkedList
+    let a = [1, 2, 3].toSinglyLinkedList
+    assert $a == $a.copy
+  result = initSinglyLinkedList[T]()
+  for x in L:
+    result.append(x)
+
+proc add*[T](L1, L2: var SinglyLinkedList[T]) {.since: (1, 5).} =
+  ## Moves `L2` to the end of `L1`. Efficiency: O(1).
+  ## Note that `L2` becomes empty after the operation.
```

Comment:
needs to explain what happens for `a.add a`

Reply:
It becomes empty, with the current implementation.

Reply:
empty sucks though;
how about raising instead? ideally would be disallowed at CT (like rust) but for now raising might be better

Reply:
We can do that, but I don't know ... somehow I feel that we concentrate on a special case too much. A lot worse than empty can happen when someone starts shuffling `head`s and `tail`s, even when sticking with the API's `append`. (You can do `a.append(a.head)`, after all, which results in a list of just the first element.)
I'm just wondering how far we should intervene... I'm all for liberty :)